torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -4546,6 +4546,147 @@ Tensor cumprod_jvp(Tensor self_t, Tensor self_p, Tensor result, int dim) {
   }
 }
 
+Tensor batch_norm_jvp(
+    const Tensor& input_p, const Tensor& input_t,
+    const Tensor& weight_p, const Tensor& weight_t,
+    const Tensor& bias_p, const Tensor& bias_t,
+    const c10::optional<Tensor>& running_mean,
+    const c10::optional<Tensor>& running_var,
+    const Tensor& saved_mean, const Tensor& saved_invstd,
+    bool train,
+    double eps) {
+  auto dims = std::vector<int64_t>{};
+  auto view_size = input_t.sizes().vec();
+  int64_t numel = 1;
+  for (const auto dim : c10::irange(view_size.size())) {
+    if (dim != 1) {
+      numel *= input_t.size(dim);
+      view_size[dim] = 1;
+      dims.push_back(dim);
+    }
+  }
+  Tensor mean_p;
+  Tensor invstd_p;
+  Tensor result_t;
+  if (train) {
+    mean_p = saved_mean.view(view_size);
+    invstd_p = saved_invstd.view(view_size);
+    auto invstd_t = -invstd_p.pow(3) * ((input_t - input_t.mean(dims, true)) * (input_p - mean_p)).sum(dims, true) / numel;
```

Comment:
There are quite a few things that could be done inplace here.
I think we want to do these. Keeping in mind the "composite compliance"

cc @zou3519 is there a doc/wiki/inline comment I can link here that states the rules of what is allowed or not?

Reply:
This: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md#compositeimplicitautograd-compliance , but it doesn't include the information about the in-place yet

Reply:
Try this as well: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorSubclassLikeUtils.h

Reply:
Thanks!
Should we have a wiki or doc note that points to these?